analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs

```diff
@@ -1082,10 +1082,110 @@ public void Main()
+                object right = {right};
+                var result = ReferenceEquals(left, right);
+                Tag("Result", result);
+                Tag("Left", left);
+                Tag("Right", right);
+                """;
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTag("Result", x => x.HasConstraint(BoolConstraint.From(expectedResult)).Should().BeTrue());
+            validator.ValidateTag("Left", x => x.AllConstraints.Select(x => x.Kind).Should().ContainSingle().Which.Should().Be(expectedConstraintLeft));
+            validator.ValidateTag("Right", x => x.AllConstraints.Select(x => x.Kind).Should().ContainSingle().Which.Should().Be(expectedConstraintRight));
+        }
+
+        [DataTestMethod]
+        [DataRow("new object()", "new object()")]
+        [DataRow("new object()", "Unknown<object>()")]
+        [DataRow("Unknown<object>()", "new object()")]
+        [DataRow("Unknown<object>()", "Unknown<object>()")]
+        [DataRow("new int?(42)", "Unknown<int?>()")]
+        [DataRow("(int?)42", "(int?)42")]
+        [DataRow("(int?)42", "(int?)0")]
+        public void Invocation_ReferenceEquals_DoesNotLearnResult(string left, string right)
+        {
+            var code = $"""
+                object left = {left};
+                object right = {right};
+                var result = ReferenceEquals(left, right);
+                Tag("Result", result);
+                """;
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTag("Result", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("null", "arg", "object")]
+        [DataRow("null", "arg", "int?")]
+        [DataRow("arg", "null", "object")]
+        [DataRow("arg", "null", "int?")]
+        public void Invocation_ReferenceEquals_SplitsToBothResults(string left, string right, string argType)
+        {
+            var code = $"""
+                var result = ReferenceEquals({left}, {right});
+                Tag("End");
+                """;
+            var validator = SETestContext.CreateCS(code, $", {argType} arg").Validator;
+            var result = validator.Symbol("result");
+            var arg = validator.Symbol("arg");
+            validator.TagStates("End").Should().SatisfyRespectively(
+                x =>
+                {
+                    x[result].HasConstraint(BoolConstraint.True).Should().BeTrue();
+                    x[arg].HasConstraint(ObjectConstraint.Null).Should().BeTrue();
+                },
+                x =>
+                {
+                    x[result].HasConstraint(BoolConstraint.False).Should().BeTrue();
+                    x[arg].HasConstraint(ObjectConstraint.NotNull).Should().BeTrue();
+                });
+        }
+
+        [TestMethod]
+        public void Invocation_ReferenceEquals_CustomSignatures_NotSupported()
```

Comment:
Educational: what do you mean here by `NotSupported`?

From the way it is implemented (using `invocation.TargetMethod.Is(KnownType.System_Object, ...` instead of the cheaper `invocation.TargetMethod.Name == nameof(object.ReferenceEquals)`) it seems to me that the code woulc correctly identify the local `ReferenceEquals` as being different than the `object.ReferenceEquals`.

As a result of that, it should not learn `ObjectConstraint.True` on `argsX`, yet it should learn `ObjectConstraint.NotNull` on them, because they are all `bool` and as such they can't be null.

So the assertion seems in line with the correct behavior.
What am I missing here?

If "unsupported" refers to the fact that the following assertion fails (behavior which is incorrect for me):
```
SETestContext.CreateCS($"""
    object left = null;
    object right = null;
    var result = ReferenceEquals(left, right);
    Tag("Result", result);
    Tag("Left", left);
    Tag("Right", right);

    bool ReferenceEquals(object o1, object o2) => false;
    """)
    .Validator
    .ValidateTag("Result", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
```
shouldn't the test rather ensure that the symbols of `left` and `right` passed to `ReferenceEquals` are not constrained, rather than asserting on the result of the custom `ReferenceEquals` implementations.

Reply:
Not supported means, that we will not simulate the correct behavior. Whatever "correct" means.

For `object.RefernceEquals`, we are sure how it behaves.
For user-defined `ReferenceEquals`, while we could assume the behavior based on the name, we just do not. We don't trust users.

So we don't learn anything at all from that invocation. The `NotNull` is learned only because it's `bool` type